neqo-crypto/src/lib.rs

```diff
@@ -107,24 +107,27 @@ fn version_check() {
 /// Initialize NSS.  This only executes the initialization routines once, so if there is any chance
 /// that
 ///
-/// # Panics
+/// # Errors
 ///
 /// When NSS initialization fails.
-pub fn init() {
+pub fn init() -> Res<()> {
     // Set time zero.
     time::init();
-    _ = INITIALIZED.get_or_init(|| {
+    let res = INITIALIZED.get_or_init(|| {
         version_check();
         if already_initialized() {
-            return NssLoaded::External;
+            return Ok(NssLoaded::External);
         }
 
-        secstatus_to_res(unsafe { nss::NSS_NoDB_Init(null()) }).expect("NSS_NoDB_Init failed");
-        secstatus_to_res(unsafe { nss::NSS_SetDomesticPolicy() })
-            .expect("NSS_SetDomesticPolicy failed");
+        secstatus_to_res(unsafe { nss::NSS_NoDB_Init(null()) })?;
+        secstatus_to_res(unsafe { nss::NSS_SetDomesticPolicy() })?;
 
-        NssLoaded::NoDb
+        Ok(NssLoaded::NoDb)
     });
+    match res {
+        Ok(_) => Ok(()),
+        Err(e) => Err(e.clone()),
+    }
```

Comment:
I think that you can do this instead:

```suggestion
    res.clone()
```

Reply:
That actually doesn't work:
```
error[E0308]: mismatched types
   --> neqo-crypto/src/lib.rs:127:5
    |
113 | pub fn init() -> Res<()> {
    |                  ------- expected `std::result::Result<(), err::Error>` because of return type
...
127 |     res.clone()
    |     ^^^^^^^^^^^ expected `Result<(), Error>`, found `&Result<NssLoaded, Error>`
    |
    = note:   expected enum `std::result::Result<(), _>`
            found reference `&std::result::Result<NssLoaded, _>`

```

Reply:
Maybe `res.map(|_| ()).clone()` @larseggert?

Reply:
Wouldn't this map all errors to not-errors? (Am still a Rust newbie.)

Reply:
No, it will only map the `Ok` variants. The `Err` enum variants are untouched.

> Maps a Result<T, E> to Result<U, E> by applying a function to a contained [Ok](https://doc.rust-lang.org/std/result/enum.Result.html#variant.Ok) value, leaving an [Err](https://doc.rust-lang.org/std/result/enum.Result.html#variant.Err) value untouched.

https://doc.rust-lang.org/std/result/enum.Result.html#method.map

Reply:
> Wouldn't this map all errors to not-errors? (Am still a Rust newbie.)

Mapping all errors to not-errors would happen when e.g. doing:

``` rust
res.and_then(|_res| Ok(()))
```

----

In more technical terms, `Result::map` is a functor transition.

``` rust
    pub fn map<U, F: FnOnce(T) -> U>(self, op: F) -> Result<U, E> {
```

`Result::and_then` is a monad transition.

``` rust
    pub fn and_then<U, F: FnOnce(T) -> Result<U, E>>(self, op: F) -> Result<U, E> {
```

Reply:
Thanks for the explanation! That didn't quite work, but 
```rs
res.as_ref().map(|_| ()).map_err(Clone::clone)
```
did.